tools/extract_embeddings.lua

```diff
@@ -40,7 +40,9 @@ local function main()
   end
   local dicts = checkpoint.dicts
   local encoder = onmt.Factory.loadEncoder(checkpoint.models.encoder)
-  local decoder = onmt.Factory.loadDecoder(checkpoint.models.decoder)
+  if checkpoint.models.decoder then
+    local decoder = onmt.Factory.loadDecoder(checkpoint.models.decoder)
```

Comment:
This variable will not be defined outside of the `if` scope. I think you can declare it just before line 59.